### PR TITLE
[WIP] Work on npc.c

### PR DESCRIPTION
**Pull request by pixel-stuck**
Opening this now for feedback on anything I might have missed or glossed over. This is still a WIP, so not all functions I'm working on are done/included.

### CODE REVIEW

File: include/common_structs.h [hunk 1]
```diff
@@ -102,7 +102,12 @@ typedef struct BlurBuffer {
     /* 0xA4 */ f32 z[20];
 } BlurBuffer; // size = 0xF4
 
-typedef s32 Palette16[16]; // size = 0x40
+typedef s16 Palette16[16]; // size = 0x20
+
+typedef union {
+    u16 h;
+    u32 w;
+} test;
```
Comment (pixel-stuck) [inline include/common_structs.h:110]: feedback on the name would be appreciated

File: src/npc.c [hunk 2]
```diff
@@ -37,45 +53,480 @@ void npc_iter_no_op(void) {
...
+extern s16 D_8010C97A;
+
+void npc_do_world_collision(Npc* npc) {
+    f32 temp_f0;
+    s32 phi_v0;
+    f32 temp_x;
+    f32 temp_y;
+    f32 temp_z;
+
+    if (npc->flags & 0x40) {
```
Comment (ethteck) [inline src/npc.c:255]: Not really an official project guidline, but I think this would be easier to read with some empty line separating the conditional blocks at least

File: src/npc.c [hunk 3]
```diff
@@ -37,45 +53,480 @@ void npc_iter_no_op(void) {
...
+                    } else if (phi_f20 > 10000.0f) {
+                        phi_f20 = 10000.0f;
+                    }
+
+                    task->distance = -phi_f20;
+                    task->appendGfxArg = npc;
+                    task->appendGfx = &npc_appendGfx;
+                    task->renderMode = npc->renderMode;
+                    if (npc->flags & 0x800000) {
+                        u8 r;
```
Comment (ethteck) [inline src/npc.c:506]: I think here it'd look nicer to do `u8 r, g, b, a;`

File: src/npc.c [hunk 4]
```diff
@@ -181,14 +660,55 @@ void update_npc_blur(Npc* npc) {
...
+                        npc->spriteInstanceID = func_802DE0EC(npc->currentAnim.w, npc->extraAnimList);
+                    } else {
+                        npc->spriteInstanceID = func_802DE0EC(npc->currentAnim.w | 0x80000000, npc->extraAnimList);
+                    }
+                }
+                if (!(npc->flags & 0x40000000)) {
+                    if (!(npc->flags & 0x1000000) && (npc->unk_B4 != 0)) {
+                        npc->spritePaletteList = func_802DEA40(npc->currentAnim.h);
+                        for (npc->paletteCount = 0; npc->spritePaletteList[npc->paletteCount] != -1; npc->paletteCount++) {
+
```
Comment (ethteck) [inline src/npc.c:695]: I think a comment would be nice here mentioning why there's an empty loop / what's happening here

File: src/npc.c (hunk 2 above)
Comment (ethteck) [inline src/npc.c:255]: Not really an official project guidline, but I think this would be easier to read with some empty line separating the conditional blocks at least

File: src/npc.c (hunk 4 above)
Comment (ethteck) [inline src/npc.c:695]: I think a comment would be nice here mentioning why there's an empty loop / what's happening here

File: src/npc.c (hunk 2 above)
Comment (ethteck) [inline src/npc.c:255]: Not really an official project guidline, but I think this would be easier to read with some empty line separating the conditional blocks at least

File: src/npc.c [hunk 5]
```diff
@@ -224,49 +806,261 @@ INCLUDE_ASM(s32, "npc", func_8003BA60);
 
 INCLUDE_ASM(s32, "npc", func_8003BED8);
 
-INCLUDE_ASM(s32, "npc", func_8003C3D8);
+void func_8003C3D8(Npc* npc, s32 idx, s8 decorationType) {
+
```
Comment (ethteck) [inline src/npc.c:810]: unnecessary empty line

File: src/npc.c [hunk 6]
```diff
@@ -224,49 +806,261 @@ INCLUDE_ASM(s32, "npc", func_8003BA60);
...
+                break;
+            case 6:
+                func_8003CCB0(npc, i);
+                break;
+        }
+    }
+}
+
+void func_8003C53C(Npc* npc, s32 idx) {
+
```
Comment (ethteck) [inline src/npc.c:852]: here too

File: src/npc.c [hunk 7]
```diff
@@ -224,49 +806,261 @@ INCLUDE_ASM(s32, "npc", func_8003BA60);
...
+    switch (npc->decorationInitialised[idx]) {
+        case 0:
+            func_80070EB0(2, npc->pos.x, npc->pos.y, npc->pos.z, 1.0f, &npc->decorations[idx]);
+            npc->decorationInitialised[idx] = 1;
+            break;
+        case 1:
+            break;
+        default:
+            return;
+
```
Comment (ethteck) [inline src/npc.c:918]: here3

File: src/npc.c [hunk 8]
```diff
@@ -224,49 +806,261 @@ INCLUDE_ASM(s32, "npc", func_8003BA60);
...
+            } else {
+                fx_sweat(0, npc->pos.x, npc->pos.y + npc->collisionHeight, npc->pos.z, 5.0f, -45.0f, 0x14);
+            }
+            npc->decorationUnk[idx] = 10;
+            npc->decorationInitialised[idx] = 1;
+            break;
+        case 1:
+            if (npc->decorationUnk[idx] != 0) {
+                npc->decorationUnk[idx]--;
+                return;
```
Comment (ethteck) [inline src/npc.c:948]: I think an else would be better than a return here

File: src/npc.c [hunk 9]
```diff
@@ -224,49 +806,261 @@ INCLUDE_ASM(s32, "npc", func_8003BA60);
...
-INCLUDE_ASM(s32, "npc", func_8003C668);
+void func_8003C8B4(Npc* npc, s32 idx) {
+    EffectInstanceData* instanceData;
 
-INCLUDE_ASM(s32, "npc", func_8003C78C);
+    switch (npc->decorationInitialised[idx]) {
+        case 0:
+            func_80070A30(0, npc->pos.x, npc->pos.y + npc->collisionHeight, npc->pos.z, 20.0f, 3, &npc->decorations[idx]);
+            npc->decorationInitialised[idx] = 1;
+            return;
```
Comment (ethteck) [inline src/npc.c:965]: break instead of return? and below

File: src/npc.c [hunk 10]
```diff
@@ -322,17 +1116,207 @@ Npc* npc_find_near_simple(f32 x, f32 y, f32 z, f32 radius) {
...
+        }
+        func_80072350(3, x, y, z,  13.0f, 10.0f, 1.0f, 5, 30);
+    } else if (arg1 != 0) {
+        if (D_80077C14++ >= 4) {
+            D_80077C14 = 0;
+            if (phi_a2 == 0) {
+                sin_cos_rad((clamp_angle(-npc->yaw) * TAU) / 360.0f, &subroutine_argA, &subroutine_argB);
+                fx_walk_normal(0, npc->pos.x + (npc->collisionRadius * subroutine_argA * 0.2f), npc->pos.y + 1.5f,
+                               npc->pos.z + (npc->collisionRadius * subroutine_argB * 0.2f), subroutine_argA, subroutine_argB);
+                return;
```
Comment (ethteck) [inline src/npc.c:1302]: else instead of return imo

File: src/npc.c (hunk 9 above)
Comment (pixel-stuck) [reply src/npc.c:965]: there's more to this function (setting the instanceData stuff) so just break is functionally different. We don't want to set the instanceData members if it was one of these specific cases.

File: src/npc.c (hunk 4 above)
Comment (pixel-stuck) [reply src/npc.c:695]: do you think a while loop might be better here?
```C
npc->palleteCount = 0;
while (npc->spritePaletteList[npc->paletteCount] != -1) {
  npc->paletteCount++;
}
```

File: Jenkinsfile [hunk 11]
```diff
@@ -49,9 +49,9 @@ pipeline {
             }
         }
     }
-    post {
-        always {
-            cleanWs()
-        }
-    }
```
Comment (nanaian) [inline Jenkinsfile:56]: Why this change? The indentation seemed correct before

File: include/common_structs.h (hunk 1 above)
Comment (nanaian) [reply include/common_structs.h:110]: Might be better to not typedef it for now and just do `union { ... } currentAnim` on L123

File: include/variables.h [hunk 12]
```diff
@@ -212,7 +212,7 @@ extern s32 gStatsMenuIconIDs[12];
 
 
 extern s32 gGameState;
-extern s16 D_8009A604;
+extern s16 D_8009A604; // num npcs
```
Comment (nanaian) [inline include/variables.h:215]: Consider renaming the symbol

File: include/common_structs.h (hunk 1 above)
Comment (nanaian) [reply include/common_structs.h:110]: Also what's the reasoning for naming the fields `h` and `w`?

File: src/npc.h [hunk 13]
```diff
@@ -211,24 +211,58 @@ void npc_dyn_entity_draw_no_op(void);
...
+/// @returns pointer to Enemy struct, if one is found. If one is not found, a panic occurs.
 Enemy* get_enemy(NpcID npcID);
 
-s32 get_enemy_safe();
+/// Same as get_enemy(), with the exception of always returning a value if an enemy is not found.
+///
+/// @param npcID   ID of the npc bound to the desired enemy.
+/// 
+/// @returns pointer to Enemy struct, if one is found. Otherwise, NULL.
+Enemy* get_enemy_safe(s32 npcID);
```
Comment (nanaian) [inline src/npc.h:266]: love this

File: src/npc.c (hunk 4 above)
Comment (nanaian) [reply src/npc.c:692]: If it matches, I'd prefer a while loop personally

File: Jenkinsfile (hunk 11 above)
Comment (pixel-stuck) [reply Jenkinsfile:56]: We were testing something on the CI and it potentially got messed up

File: include/common_structs.h (hunk 1 above)
Comment (pixel-stuck) [reply include/common_structs.h:110]: sounds like a good idea to just do an anonymous union. As for the field names; "halfword" and "word" are the idea behind them

Comment (ethteck) [reply include/common_structs.h:110]: oh I was thinking height and width. I think we might want to go with just u32 and u16 or something
